### ISSUE TITLE
When I reload a certificate, oatpp close my ConnectionProvider

### ISSUE THREAD

**Issue by npastuh0v** [oatpp/oatpp#955]
Hi! I'm reload my certificate and I ran with a problem. We have a custom-realized class AppComponet, code of this class is given below:
`
class AppComponentCustomSSL {
private:
  std::shared_ptr<oatpp::libressl::Config>                         sslConfig_handler                      = nullptr; 
  std::shared_ptr<oatpp::network::tcp::server::ConnectionProvider> inputConnectionProvider_handler        = nullptr;  
  std::shared_ptr<oatpp::libressl::server::ConnectionProvider>     connectionProvider_handler             = nullptr;  
  std::shared_ptr<oatpp::network::ServerConnectionProvider>        lastConnectionMonitor_handler          = nullptr; 
  std::shared_ptr<oatpp::network::ConnectionProviderSwitch>        serverConnectionProviderSwitch_handler = nullptr; 
  std::shared_ptr<oatpp::web::server::HttpRouter>                  sslRouter_handler                      = nullptr; 
  std::shared_ptr<oatpp::network::ConnectionHandler>               sslConnectionHandler_handler           = nullptr; 
  std::shared_ptr<oatpp::network::Server>                          HTTPSServer_handler                    = nullptr; 
  
  std::atomic_bool HTTPS_ENABLED                                                                         = false;
  std::atomic_bool HTTPS_SERVER_IS_WORKING                                                               = false;
  std::atomic_bool SERVER_CURRENT_WORKING_STATUS                                                         = false;

  void initalizeInputConnectionProvider() {
    v_uint16 WEB_SERVER_HTTPS_PORT = 443;
    this->inputConnectionProvider_handler = oatpp::network::tcp::server::ConnectionProvider::createShared({"0.0.0.0", WEB_SERVER_HTTPS_PORT});
  }

  void initalizeServerConnectionProvider() {
    if(sslConfig_handler != nullptr) {
      sslConfig_handler.reset();
    }
    sslConfig_handler = oatpp::libressl::Config::createDefaultServerConfigShared("KEY_FILE_PATH", "CRT_FILE_PATH");
    connectionProvider_handler = oatpp::libressl::server::ConnectionProvider::createShared(sslConfig_handler, this->inputConnectionProvider_handler);
    return;
  }

  void initalizeMonitorWithInternalConnectionProvider() {
    if(this->inputConnectionProvider_handler == nullptr) {
      this->initalizeInputConnectionProvider();
    }
    if(this->connectionProvider_handler == nullptr) {
      this->initalizeServerConnectionProvider();
    }
    std::shared_ptr<oatpp::network::monitor::ConnectionMonitor> monitor_tmp = std::make_shared<oatpp::network::monitor::ConnectionMonitor>(connectionProvider_handler);
    monitor_tmp->addMetricsChecker( std::make_shared<oatpp::network::monitor::ConnectionMaxAgeChecker>( std::chrono::seconds(3600) ) );
    this->lastConnectionMonitor_handler = monitor_tmp;
    monitor_tmp.reset();
    return;
  }

  void initalizeServerConnectionProviderSwitch() {
    this->serverConnectionProviderSwitch_handler = std::make_shared<oatpp::network::ConnectionProviderSwitch>(this->connectionProvider_handler);
  }

  void initalizeSSLRouter() {
    this->sslRouter_handler = oatpp::web::server::HttpRouter::createShared();
  }

  void initalizeSSLConnectionHandler() {
    std::shared_ptr<oatpp::web::server::HttpConnectionHandler> temp_ptr = oatpp::web::server::HttpConnectionHandler::createShared(this->sslRouter_handler);
    temp_ptr->addRequestInterceptor(std::make_shared<oatpp::web::server::interceptor::AllowOptionsGlobal>());
    temp_ptr->addResponseInterceptor(std::make_shared<oatpp::web::server::interceptor::AllowCorsGlobal>("*", "GET, POST, OPTIONS", "DNT, User-Agent, X-Requested-With, If-Modified-Since, Cache-Control, Content-Type, Range, Authorization, X-Vsaas-Api-Key"));
    this->sslConnectionHandler_handler = temp_ptr;
    temp_ptr.reset();
  }

public:
  void initalizeComponents() {
    this->initalizeInputConnectionProvider                 ();
    this->initalizeServerConnectionProvider                ();
    this->initalizeMonitorWithInternalConnectionProvider   ();
    this->initalizeServerConnectionProviderSwitch          ();
    this->initalizeSSLRouter                               ();
    this->initalizeSSLConnectionHandler                    ();
  }

  void destroyComponents() {
    this->resetInputConnectionProvider_handler             ();
    this->resetSSLConfig_handler                           ();
    this->resetConnectionProvider_handler                  ();
    this->resetLastConnectionMonitor_handler               ();
    this->resetServerConnectionProviderSwitch_handler      ();
    this->resetSSLRouter_handler                           ();
    this->resetSSLConnectionHandler_handler                ();
  }

  AppComponentCustomSSL() { }
  
  ~AppComponentCustomSSL() {
    this->destroyComponents();
  }

  std::shared_ptr<oatpp::network::tcp::server::ConnectionProvider> *getInputServerConnectionProvider() {
    return &this->inputConnectionProvider_handler;
  }
  
  std::shared_ptr<oatpp::network::ServerConnectionProvider> *getLastMonitorWithInternalConnectionProvider()  {
    return &this->lastConnectionMonitor_handler;
  }

  std::shared_ptr<oatpp::libressl::server::ConnectionProvider> *getServerConnectionProvider()  {
    return &this->connectionProvider_handler;
  }

  std::shared_ptr<oatpp::network::ConnectionProviderSwitch> *getServerConnectionProviderSwitch()  {
    return &this->serverConnectionProviderSwitch_handler;
  }

  std::shared_ptr<oatpp::web::server::HttpRouter> *getSSLRouter()  {
    return &this->sslRouter_handler;
  }

  std::shared_ptr<oatpp::network::ConnectionHandler> *getSSLConnectionHandler()  {
    return &this->sslConnectionHandler_handler;
  }

  void resetInputConnectionProvider_handler() {
    if(this->inputConnectionProvider_handler != nullptr) {
      this->inputConnectionProvider_handler.reset();
      this->inputConnectionProvider_handler = nullptr;
    }
  }

  void resetSSLConfig_handler() {
    if(this->sslConfig_handler != nullptr) {
      this->sslConfig_handler.reset();
      this->sslConfig_handler = nullptr;
    }
  }

  void resetConnectionProvider_handler() {
    if(this->connectionProvider_handler != nullptr) {
      this->connectionProvider_handler.reset();
      this->connectionProvider_handler = nullptr;
    }
  }

  void resetLastConnectionMonitor_handler() {
    if(this->lastConnectionMonitor_handler != nullptr) {
      this->lastConnectionMonitor_handler.reset();
      this->lastConnectionMonitor_handler = nullptr;
    }
  }

  void resetServerConnectionProviderSwitch_handler() {
    if(this->serverConnectionProviderSwitch_handler != nullptr) {
      this->serverConnectionProviderSwitch_handler.reset();
      this->serverConnectionProviderSwitch_handler = nullptr;
    }
  }

  void resetSSLRouter_handler() {
    if(this->sslRouter_handler != nullptr) {
      this->sslRouter_handler.reset();
      this->sslRouter_handler = nullptr;
    }
  }

  void resetSSLConnectionHandler_handler() {
    if(this->sslConnectionHandler_handler != nullptr) {
      this->sslConnectionHandler_handler.reset();
      this->sslConnectionHandler_handler = nullptr;
    }
  }
  void reloadCert() {
      v_uint16 WEB_SERVER_HTTPS_PORT = MSConfig->getHttpsPort();
      try {
        this->resetConnectionProvider_handler();
        this->initalizeServerConnectionProvider();
        std::shared_ptr<oatpp::libressl::server::ConnectionProvider> provider_temp = oatpp::libressl::server::ConnectionProvider::createShared(this->sslConfig_handler, this->inputConnectionProvider_handler);
        this->serverConnectionProviderSwitch_handler->resetProvider(provider_temp);
        this->connectionProvider_handler.reset();
        this->connectionProvider_handler = provider_temp;
        provider_temp.reset();
        this->lastConnectionMonitor_handler.reset();
        this->initalizeMonitorWithInternalConnectionProvider();
        std::cout << "Cert Reloaded Successfully!!!!" << std::endl;
        SSL_CERTIFICATE_RELOAD_ERROR = false;
        return;
      } catch(std::runtime_error &e) {
        SSL_CERTIFICATE_RELOAD_ERROR = true;
        std::cout << "Cert Reloading ERROR!!!!" << std::endl;
        return;
      }
  }
  void pauseServer() {
    this->serverConnectionProviderSwitch_handler->stop();
    this->resetLastConnectionMonitor_handler();
    this->resetConnectionProvider_handler();
    this->resetInputConnectionProvider_handler();
    this->resetSSLConfig_handler();
  }

  void resumeServer() {
    this->initalizeInputConnectionProvider();
    this->initalizeServerConnectionProvider();
    this->initalizeMonitorWithInternalConnectionProvider();
    this->serverConnectionProviderSwitch_handler->resetProvider(this->connectionProvider_handler);
  }
};
`
Example of code, which we uses instance of this class, is given below:
`
AppComponentCustomSSL sslServiceCustom;
std::atomic_bool WEB_SERVER_HTTPS_START;
while(1) {
    // ...
    
    sslServiceCustom.initalizeComponents();
    std::shared_ptr<oatpp::web::server::HttpRouter> *router = componentsSSL.getSSLRouter();
    std::shared_ptr<oatpp::network::ConnectionHandler> *handler = componentsSSL.getSSLConnectionHandler();
    std::shared_ptr<oatpp::network::ConnectionProviderSwitch> *provider = componentsSSL.getServerConnectionProviderSwitch();
    
    router->get()->addController(std::make_shared<Controller>());
    std::shared_ptr<oatpp::network::Server> oatppCustomSSLServer = oatpp::network::Server::createShared(*provider, *handler);
    std::thread oatppCustomSSLServerThread([&oatppCustomSSLServer]{
        try{
            WEB_SERVER_HTTPS_START = true;
            server->run();
            WEB_SERVER_HTTPS_START = false;
        } catch(std::runtime_error &e){
            std::cout << "Oops! An error occoured when we runs a HTTPS server!" << std::endl;
            WEB_SERVER_HTTPS_START = false;
        }
    });
    oatppCustomSSLServerThread.detach();

    std::cout << "Sleeping #1" << std::endl;
    sleep(30);
    std::cout << "Certificate reloading... Challenge № 1" << std::endl;
    sslServiceCustom.reloadCert();
    std::cout << "Sleeping #2" << std::endl;
    sleep(30);
    std::cout << "Certificate reloading... Challenge № 2" << std::endl;
    sslServiceCustom.reloadCert();
    std::cout << "Sleeping #3" << std::endl;
    sleep(30);

    // ...
    sslServiceCustom.destroyComponents();
}
`
The crux of the problem is that before the first call to the **reloadCert()** method, the server is working great. After calling this method, the server is no longer available, and even calling the netstat utility with the "-lntup" keys does not display it. Please tell me, what could be the reason for this behavior? 
Thanks!

**Comment by lganzzzo** [comment]
Hello @npastuh0v ,

I believe this code isn't working. I can't be sure because I have trouble reading it.
Things here are way too overcomplicated.

Please simplify it and most probably you'll fix it in the process.

**Comment by npastuh0v** [comment]
Okay! I simplified this class and moved on to using "ConnectionProviderSwitch". But it doesn't work again. My code:
```
class AppComponentCustomSSL {
//...
// definition of ConnectionProviderSwitch
static void reloadCert() {
auto config = oatpp::libressl::Config::createDefaultServerConfigShared(MSConfig->getParam("PEM_FILE").c_str(), MSConfig->getParam("CRT_FILE").c_str());
serverConnectionProviderSwitch->resetProvider(connectionProvider);
};
// ...
};
```
What I do wrong? Why it doesn`t works? Thanks for help
P.S.: Callback for libressl looks like this:
```
```
void callback_sighup(int signal) {
    if(signal == SIGHUP) {
        // SIGHUP received!!
        AppComponentCustomSSL::reloadCert();
    }
}
```
```
oatpp::libressl::Callbacks::setDefaultCallbacks();
signal(SIGHUP, callback_sighup);
```
